tensorflow-keras/src/main/java/org/tensorflow/keras/utils/SymbolicShapeDict.java

```diff
@@ -0,0 +1,58 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================*/
+package org.tensorflow.keras.utils;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Utility class that contains a dictionary of Symbols. The map is updated with symbols and their
+ * dimension values. Used in creating matching shapes that depend on other shapes' dimensions.
+ */
+public class SymbolicShapeDict {
+
+  /** Dictionary containing the sybmols and their associated dimensions sizes. */
+  private final Map<String, Long> map = new HashMap<>();
+
+  /**
+   * Add a symbol with its associated dimensions size.
+   *
+   * @param symbol the symbol name
+   * @param size the shape dimension to associate with the symbol
+   */
+  public void put(String symbol, Long size) {
+    this.map.put(symbol, size);
```

Comment:
Do we want the ability to overwrite values that already exist? If so, should we also have a remove method?

Reply:
I could change the name to "add", then have the method remove the old one first.

Reply:
Well it'll overwrite them at the moment, but it's more from a semantics perspective, do we need the ability to overwrite symbol names, and if we have it will it cause trouble.

Reply:
`SymbolicShapes` define a set of rules for shapes that a set of `Operand`s have in common. Once created, they are used to create `AssertThat`'s to make sure the rules are followed among the `Operand`s. This is done within one method, as method variables, so currently, there is no chance for the rules to be replaced later.

Usually the symbols are created as a group within  a `Metric`, primarily the ConfusionMatrix type metrics (AUC), 
After they are set up, then the `AssertThat`'s are created and the `SymbolicShapes` are no longer used.

Reply:
Also, the only place that I have currently found where this used is in the `AUC` (Area Under Curve) Metric. We could implement the `Map` and Symbolic Shape classes inside of the `AUC` class.

Reply:
That sounds good.

Reply:
SymbolicShapes is not need until I check in Metrics, so I removed it from this PR.